src/Confluent.Kafka/Producer.cs

```diff
@@ -432,145 +576,130 @@ public int Poll()
-        ///     Warning: blockIfQueueFull is set to true, background polling is 
-        ///     disabled and Poll is not being called in another thread, this
-        ///     will block indefinitely.
-        /// </param>
         /// <returns>
         ///     A Task which will complete with the corresponding delivery report
         ///     for this request.
         /// </returns>
         /// <remarks>
+        ///     The partition the message is produced to is determined using the configured partitioner.
+        ///     
+        ///     Blocks if the send queue is full. Warning: if background polling is disabled and Poll is
+        ///     not being called in another thread, this will block indefinitely.
+        ///     
         ///     If you require strict ordering of delivery reports to be maintained, 
         ///     you should use a variant of ProduceAsync that takes an IDeliveryHandler
         ///     parameter, not a variant that returns a Task&lt;Message&gt; because 
         ///     Tasks are completed on arbitrary thread pool threads and can 
         ///     be executed out of order.
         /// </remarks>
-        public Task<Message> ProduceAsync(string topic, byte[] key, int keyOffset, int keyLength, byte[] val, int valOffset, int valLength, int partition, bool blockIfQueueFull)
-            => Produce(topic, val, valOffset, valLength, key, keyOffset, keyLength, null, partition, blockIfQueueFull);
+        public Task<Message> ProduceAsync(string topic, byte[] key, byte[] val)
+            => Produce(topic, val, 0, val?.Length ?? 0, key, 0, key?.Length ?? 0, null, RD_KAFKA_PARTITION_UA, true);
 
         /// <summary>
         ///     Asynchronously send a single message to the broker.
-        ///     Refer to <see cref="ProduceAsync(string, byte[], int, int, byte[], int, int, int, bool)" />
-        ///     for more information.
         /// </summary>
-        public Task<Message> ProduceAsync(string topic, byte[] key, int keyOffset, int keyLength, byte[] val, int valOffset, int valLength, bool blockIfQueueFull)
-            => Produce(topic, val, valOffset, valLength, key, keyOffset, keyLength, null, RD_KAFKA_PARTITION_UA, blockIfQueueFull);
-
+        /// <param name="produceRecord">
+        ///     The record informations.
+        /// </param>
+        /// <returns>
+        ///     A Task which will complete with the corresponding delivery report
+        ///     for this request.
+        /// </returns>
+        /// <remarks>
+        ///     If you require strict ordering of delivery reports to be maintained, 
+        ///     you should use a variant of ProduceAsync that takes an IDeliveryHandler
+        ///     parameter, not a variant that returns a Task&lt;Message&gt; because 
+        ///     Tasks are completed on arbitrary thread pool threads and can 
+        ///     be executed out of order.
+        /// </remarks>
+        public Task<Message> ProduceAsync(ProduceRecord produceRecord)
```

Comment:
on the non-generic producer, i'm pretty sure I prefer not having a `ProduceRecord` overload. We can probably just get rid of a lot of the methods and keep the all inclusive one.